### PR TITLE
[Tts/Stt] Use thread lock only when add and remove event handler

### PR DESCRIPTION
**Pull request by stom-hwang**
### Description of Change ###
This patch creates the thread lock for event handlers and removes other useless thread locks.

### API Changes ###

 - No ACR

### CODE REVIEW

File: src/Tizen.Uix.Stt/Tizen.Uix.Stt/SttClient.cs [hunk 1]
```diff
@@ -389,7 +393,11 @@ public SttClient()
                             if (data != null && msg != null)
                             {
                                 RecognitionResultEventArgs args = new RecognitionResultEventArgs(e, data, dataCount, Marshal.PtrToStringAnsi(msg));
-                                _recognitionResult?.Invoke(this, args);
+
+                                lock (_recognitionResultLock)
+                                {
+                                    _recognitionResult?.Invoke(this, args);
+                                }
```
Comment (myroot) [inline src/Tizen.Uix.Stt/Tizen.Uix.Stt/SttClient.cs:400]: No need lock here, `_recognitionResult?.`  it guarantee thread safe
```suggestion
                                _recognitionResult?.Invoke(this, args);
```

File: src/Tizen.Uix.Stt/Tizen.Uix.Stt/SttClient.cs [hunk 2]
```diff
@@ -432,15 +440,19 @@ public SttClient()
...
                         SttClient obj = this;
                         _stateDelegate = (IntPtr handle, State previous, State current, IntPtr userData) =>
                         {
                             StateChangedEventArgs args = new StateChangedEventArgs(previous, current);
-                            _stateChanged?.Invoke(obj, args);
+
+                            lock (_stateChangedLock)
+                            {
+                                _stateChanged?.Invoke(obj, args);
+                            }
```
Comment (myroot) [inline src/Tizen.Uix.Stt/Tizen.Uix.Stt/SttClient.cs:455]: No need lock here
```suggestion
                            _stateChanged?.Invoke(obj, args);
```

File: src/Tizen.Uix.Stt/Tizen.Uix.Stt/SttClient.cs [hunk 3]
```diff
@@ -479,14 +490,18 @@ public SttClient()
...
                     {
                         _errorDelegate = (IntPtr handle, SttError reason, IntPtr userData) =>
                         {
                             ErrorOccurredEventArgs args = new ErrorOccurredEventArgs(handle, reason);
-                            _errorOccurred?.Invoke(this, args);
+
+                            lock (_errorOccurredLock)
+                            {
+                                _errorOccurred?.Invoke(this, args);
+                            }
```
Comment (myroot) [inline src/Tizen.Uix.Stt/Tizen.Uix.Stt/SttClient.cs:504]: No need lock here
```suggestion
                            _errorOccurred?.Invoke(this, args);
```

File: src/Tizen.Uix.Stt/Tizen.Uix.Stt/SttClient.cs [hunk 4]
```diff
@@ -533,7 +548,11 @@ public SttClient()
                             string previousLanguageString = Marshal.PtrToStringAnsi(previousLanguage);
                             string currentLanguageString = Marshal.PtrToStringAnsi(currentLanguage);
                             DefaultLanguageChangedEventArgs args = new DefaultLanguageChangedEventArgs(previousLanguageString, currentLanguageString);
-                            _defaultLanguageChanged?.Invoke(this, args);
+
+                            lock (_defaultLanguageChangedLock)
+                            {
+                                _defaultLanguageChanged?.Invoke(this, args);
+                            }
```
Comment (myroot) [inline src/Tizen.Uix.Stt/Tizen.Uix.Stt/SttClient.cs:555]: No need lock here
```suggestion
                            _defaultLanguageChanged?.Invoke(this, args);
```

File: src/Tizen.Uix.Stt/Tizen.Uix.Stt/SttClient.cs [hunk 5]
```diff
@@ -581,7 +600,11 @@ public SttClient()
                             string engineIdString = Marshal.PtrToStringAnsi(engineId);
                             string languageString = Marshal.PtrToStringAnsi(language);
                             EngineChangedEventArgs args = new EngineChangedEventArgs(engineIdString, languageString, supportSilence, needCredential);
-                            _engineChanged?.Invoke(this, args);
+
+                            lock (_engineChangedLock)
+                            {
+                                _engineChanged?.Invoke(this, args);
+                            }
```
Comment (myroot) [inline src/Tizen.Uix.Stt/Tizen.Uix.Stt/SttClient.cs:607]: No need lock

**Review by myroot — CHANGES_REQUESTED** [review]

**Comment by stom-hwang** [comment]
Thank you to share the information.
I will remove the thread lock from delegator invocation.

**Review by myroot — APPROVED** [review]

**Review by urmain — APPROVED** [review]
LGTM

**Review by syeonholic — APPROVED** [review]
It looks good to me.

**Pull request merged**